Payment request is submitted when auction is accepted

adelevie opened the pull request:
Fixes: https://github.com/18F/micropurchase/issues/417

TODO:
- ~~cucumber steps~~

jessieay commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,31 @@
+        purchase_type: "Software"
+      }
+    }
+    proposal = resp.body
```
ah ok so the response includes the `id` of the proposal? that's cool. we might consider asking them to return the full URL as well since that would be in line with HATEOAS https://en.wikipedia.org/wiki/HATEOAS

harrisj replied:
HATEOAS gonna HATE... oh sorry, I couldn't resist. Yes, the full URL would be great.

harrisj commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,31 @@
+class CreateCapProposal
+  def initialize(auction)
+    @host = ENV.fetch('C2_HOST', 'https://cap.18f.gov')
+    @c2_client = C2::Client.new(
```
I wonder if we should move this into an initializer so it'll blow up immediately if those ENV vars are not found (right now it'll just error out the jobs). But I might be overthinking this.

jessieay commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,90 @@
+class CreateCapProposal
+  def initialize(auction)
+    if !auction.is_a?(Presenter::Auction)
+      fail "an instance of Auction::Presenter must be passed to CreateCapProposal::initialize"
```
it seems like we shouldn't have to type check like this in our `initialize` methods

adelevie replied:
Fair enough. I'll remove.

jessieay commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,90 @@
+    @auction = auction
+  end
+
+  def configure_c2_client!
```
I think I agree w @harrisj 's earlier comments: can we do this in `config/initializers` ?

jessieay commented on spec/models/create_cap_proposal_spec.rb:
```diff
@@ -0,0 +1,35 @@
+require 'rails_helper'
+
+RSpec.describe CreateCapProposal do
```
these classes look like they belong in `app/services`

jessieay commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,90 @@
+  end
+
+  def c2_client
+    @c2_client ||= configure_c2_client!
```
since `configure_c2_client!` does not destroy data, I would expect it to be called `configure_c2_client` (save the bangs for destructive behaviors)

jessieay commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,90 @@
+
+  def c2_proposal_hash
+    {
+      office: 'DC',
```
can we put `DC` in a constant? feels like a magical value here

jessieay commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,90 @@
+      quantity: 1,
+      recurring: false,
+      date_requested: date_requested,
+      urgency: 20,
```
can we put this in a constant to reveal naming? I think we could use enum value, too :) (hashtag inside knowledge of C2)

jessieay commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,90 @@
+
+  def product_name_and_description
+    %{
+      Micropurchase for '#{@auction.title}.
```
what's the deal with the single quote here?

adelevie replied:
Typo.

jessieay commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,90 @@
+    %{
+      Micropurchase for '#{@auction.title}.
+
+      Link: https://micropurchase.18f.gov/auctions/#{@auction.id}.
```
thoughts on using a `auction_url(auction)` helper here instead so this url will work on staging as well? you can use the methods described in here to access those helpers outside of controllers/ views: https://inopinatus.org/2015/09/08/using-url-helpers-in-rails-activejob-background-jobs/

adelevie commented on app/models/vcap_application.rb:
```diff
@@ -0,0 +1,9 @@
+class VCAPApplication
+  def self.application_uris
+    vcap_application['application_uris']
```
TODO: encapsulate environment logic here (move it out of Presenter::Auction#url)

adelevie commented on app/models/update_auction.rb:
```diff
@@ -1,12 +1,27 @@
 class UpdateAuction < Struct.new(:auction, :params)
```
consider moving to services folder

adelevie commented on app/models/update_auction.rb:
```diff
@@ -1,12 +1,27 @@
+    auction.assign_attributes(attributes)
+
+    if should_create_cap_proposal?
+      cap_proposal_url = CreateCapProposal.new(auction).perform
```
make sure this is a presenter

adelevie commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,92 @@
+    @auction = auction
+  end
+
+  def configure_c2_client
```
make this private

adelevie commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,92 @@
+    @c2_client
+  end
+
+  def c2_client
```
reorder methods to the order they are used.

adelevie commented on app/models/create_cap_proposal.rb:
```diff
@@ -0,0 +1,92 @@
+  end
+
+  def date_requested
+    Date.today.iso8601
```
consider Date.current (instead of Date.today) -- current is in the timezone of the application ... research this.

adelevie commented on spec/models/create_cap_proposal_spec.rb:
```diff
@@ -0,0 +1,40 @@
+  let(:create_cap_proposal) { CreateCapProposal.new(auction_presenter) }
+  let(:c2_proposal_hash) { create_cap_proposal.c2_proposal_hash }
+
+  describe '#c2_proposal_hash' do
```
make #c2_proposal_hash private, and test #perform instead. Do something like

```
expect(c2_client).to receive(:post).with(proposals', c2_proposal_hash_stub).and_return(mocked_response)
```

....

adelevie commented on spec/models/update_auction_spec.rb:
```diff
@@ -0,0 +1,112 @@
+
+      it 'sets cap_proposal_url' do
+        expect { updater.perform }.to change { auction.cap_proposal_url }
+        expect(auction.cap_proposal_url).to be_a(String)
```
I can remove this!

adelevie commented on spec/models/update_auction_spec.rb:
```diff
@@ -0,0 +1,112 @@
+
+      it 'does not set cap_proposal_url' do
+        expect { updater.perform }.to_not change { auction.cap_proposal_url }
+        expect(auction.cap_proposal_url).to_not be_a(String)
```
I can remove this!

adelevie commented on spec/models/update_auction_spec.rb:
```diff
@@ -0,0 +1,112 @@
+    end
+  end
+
+  describe '#should_create_cap_proposal?' do
```
make this private and remove these tests.

adelevie commented on spec/support/url_matcher.rb:
```diff
@@ -0,0 +1,7 @@
+RSpec::Matchers.define :be_url do |expected|
```
holy spacing

jessieay commented:
bunch of screenshots from failed cucumber specs snuck in here. also sorry for merge conflicts! The helpers have been renamed

adelevie commented:
removed screenshots, rebased (way less painful than I thought), and now working on:

- making FakeCapAPI
- finishing the last expectation in the cucumber spec (something like "Then I should see a CAP Proposal URL")

gemfarmer commented on app/models/presenter/auction.rb:
```diff
@@ -0,0 +1,196 @@
+# require 'action_view'
```
should this be deleted instead of commented out?

harrisj replied:
I think this is a duplicate file from a rebase after Jessie moving things around in #506. It can be deleted entirely.

gemfarmer commented on app/presenters/auction_presenter.rb:
```diff
@@ -132,6 +134,15 @@ def human_start_time
 
+  def url
+    root_url = if Rails.env.development? || Rails.env.test?
+                 ENV['ROOT_URL']
```
looks like some extra whitespace. nit picky, I know

adelevie replied:
Code Climate/Rubucop demanded it. Bow before your static code analysis overlords!

gemfarmer replied:
haha 🙇

gemfarmer commented on features/step_definitions/flash_message_steps.rb:
```diff
@@ -21,9 +21,10 @@
-    expect(page).to have_content(message)
-  end
+  expect(page).to have_content(message)
+  # within("div.usa-alert.usa-alert-warning") do
```
remove?

harrisj replied:
Yes, please remove

harrisj commented on app/jobs/create_cap_proposal_job.rb:
```diff
@@ -0,0 +1,10 @@
+    auction = Auction.find(auction_id)
+    auction_presenter = AuctionPresenter.new(auction)
+    cap_proposal_url = CreateCapProposal.new(auction_presenter).perform
+    auction.update(cap_proposal_url: cap_proposal_url)
```
Could we have this be updated as part of the `perform` method? Could also define the presenter within it

adelevie replied:
I like that idea! Implementing now ...

harrisj commented on spec/presenters/auction_presenter_spec.rb:
```diff
@@ -103,6 +110,7 @@
         expect(lowest_bids.map(&:id)).to eq(ar_lowest_bids.map(&:id))
         expect(lowest_bids.map(&:bidder_id)).to eq(ar_lowest_bids.map(&:bidder_id))
       end
+
```
Get rid of this random whitespace

harrisj commented:
Looks good generally! There is one issue with an old version of the `Presenter::Auction` that seems to be in the wrong place with the rebase, but otherwise...

adelevie commented:
After lots of cleanup, I think it's ready for some quick review + merge.

harrisj commented on spec/rails_helper.rb:
```diff
@@ -1,4 +1,4 @@
-ENV['RAILS_ENV'] ||= 'test'
+ENV['RAILS_ENV'] = 'test'
```
I don't imagine there is a time when we'd run the tests outside of the 'test' environment, but this would probably make for inscrutable errors if we had a separate environment for tests on Travis vs. local machine. Is there a reason you changed this here (not saying it's wrong, just curious)

harrisj commented:
Looks good to me. Merging.
